From a38e31b385970c8c397b1b31221022b62255545f Mon Sep 17 00:00:00 2001
From: Andrzej Surdej <Andrzej_Surdej@comcast.com>
Date: Mon, 10 Jun 2024 12:18:03 +0200
Subject: [PATCH] [GStreamer] Fix negative timestamps handling (edit lists)

Bring back negative timestamps clamp on conversion to gst time
that was removed by WebRTC related backports from upstream:
95cae32e59cfdfad7cdd7139909bb47a7aed494e
The part of code comes from Edit list support patch that
is entirely reverted upstream.

This fixes YT MSE Codec Tests 84.BufUnbufSeekH264
---
 Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.h | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.h b/Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.h
index 16147380c88e..e1b67aff3247 100644
--- a/Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.h
+++ b/Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.h
@@ -81,6 +81,10 @@ uint64_t toGstUnsigned64Time(const MediaTime&);
 
 inline GstClockTime toGstClockTime(const MediaTime& mediaTime)
 {
+    if (mediaTime.isInvalid())
+        return GST_CLOCK_TIME_NONE;
+    if (mediaTime < MediaTime::zeroTime())
+        return 0;
     return static_cast<GstClockTime>(toGstUnsigned64Time(mediaTime));
 }
 
-- 
2.25.1

